src/jpparse_fold.erl

```diff
@@ -0,0 +1,107 @@
+-module(jpparse_fold).
+
+-export([foldtd/3, foldbu/3, string/1,
+         roots/1]).
+
+% Folds a json path parse tree into a string that represents a same jppath from
+% which the parse tree was originally constructed
+-spec string(tuple()) -> {ok, binary()}.
+string(Pt) ->
+    {ok, list_to_binary(foldbu(fun stringfun/3, [], Pt))}.
+
+stringfun(_Depth, {'fun',_,Args} = _Pt, Stk) ->
+    {PArgs, [A|Rest]} = lists:split(length(Args), Stk),
+    [string:join([A, "(", string:join(lists:reverse(PArgs), ","), ")"], "")
+     | Rest];
+stringfun(_Depth, {Op,_,Args} = _Pt, Stk) when Op =:= '[]'; Op =:= '{}' ->
+    {PArgs, [A|Rest]} = lists:split(length(Args), Stk),
+    {Lb,Rb} = case Op of
+                  '[]' -> {"[","]"};
+                  '{}' -> {"{","}"}
+              end,
+    [string:join([A, Lb, string:join(lists:reverse(PArgs), ","), Rb], "")
+     | Rest];
+stringfun(_Depth, {O,_,_} = _Pt, Stk) when O =:= ':'; O =:= '::'; O =:= '#' ->
+    [B,A|Rest] = Stk,
+    [string:join([A,atom_to_list(O),B], "")|Rest];
+stringfun(_Depth, Pt, Stk) when is_binary(Pt) ->
+    [binary_to_list(Pt)|Stk];
+stringfun(_Depth, Pt, Stk) when is_integer(Pt) ->
+    [integer_to_list(Pt)|Stk].
+
+% Folds a json path parse tree returning path roots
+%  as sorted array
+-spec roots(tuple()) -> {ok, list()}.
+roots(Pt) ->
+    case foldbu(fun rootsfun/3, [], Pt) of
+        {error, _} = Error -> Error;
+        Folded -> {ok, lists:usort(Folded)}
+    end.
+
+rootsfun(_Depth, {':',_,R},  Rs) when is_binary(R) -> [R|Rs];
+rootsfun(_Depth, {'::',_,R}, Rs) when is_binary(R) -> [R|Rs];
+rootsfun(_Depth, {'#',_,R},  Rs) when is_binary(R) -> [R|Rs];
+rootsfun(_Depth, {'[]',R,_}, Rs) when is_binary(R) -> [R|Rs];
+rootsfun(_Depth, {'{}',R,_}, Rs) when is_binary(R) -> [R|Rs];
+rootsfun(_Depth, _Pt, Rs) -> Rs.
+
+% Folds a jppath parsee tree applying fold function
+%  in a Top-Down walk
+-spec foldtd(Function :: fun((Depth :: integer(), SubParseTree :: any(),
+                              AccIn :: any()) -> AccOut :: any()),
+                             Acc :: any(),
+                             ParseTree :: tuple()) -> any().
+foldtd(Fun, Acc, Pt) when is_function(Fun, 3) ->
+    fold_i({td,Fun}, Acc, Pt).
+
+% Folds a jppath parsee tree applying fold function
+%  in a Boottom-Up walk
+-spec foldbu(Function :: fun((Depth :: integer(), SubParseTree :: any(),
+                              AccIn :: any()) -> AccOut :: any()),
+                             Acc :: any(),
+                             ParseTree :: tuple()) -> any().
```

Comment:
`Acc` and `ParseTree` are parameters of `foldbu` vertically aligning them with `Function` parameter is preferred!